src/utils/efd.c

```diff
@@ -44,18 +45,60 @@ int nn_efd_wait (struct nn_efd *self, int timeout)
 {
     int rc;
     struct pollfd pfd;
+    uint64_t expire;
 
-    pfd.fd = nn_efd_getfd (self);
-    pfd.events = POLLIN;
-    if (nn_slow (pfd.fd < 0))
-        return -EBADF;
-    rc = poll (&pfd, 1, timeout);
-    if (nn_slow (rc < 0 && errno == EINTR))
-        return -EINTR;
-    errno_assert (rc >= 0);
-    if (nn_slow (rc == 0))
-        return -ETIMEDOUT;
-    return 0;
+    if (timeout > 0) {
+        expire = nn_clock_ms() + timeout;
+    } else {
+        expire = timeout;
+    }
```

Comment:
It's not immediately clear what values <-1 would do. I believe these are prevented upstream, and if that's the case, it would make sense to insert an `nn_assert (timeout == -1);` here, perhaps with a comment why.